internal/migrate/versions.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package migrate
+
+import (
+	"database/sql"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+)
+
+/*
+
+Scenarios it doesn't handle properly.
+
+1. Rollback to initial state on multi-step migration.
+
+	Let's say there's a scenario where we run migration steps:
+	1. update a table schema
+	2. move files
+	3. update a table schema
+	4. update a table schema, which fails
+
+	In this case there's no easy way to rollback the moving of files.
+
+2. Undoing migrations.
+
+	Intentionally left out, because we do not gain that much from currently.
+
+3. Snapshotting the whole state.
+
+	This probably should be done by the user of this library, when there's disk-space available.
+
+4. Figuring out what the exact executed steps are.
+*/
+
+// Migration describes a migration steps
+type Migration struct {
+	Table    string
+	OnCreate Action
+	Steps    []*Step
+}
+
+// Step describes a single step in migration.
+type Step struct {
+	Description string
+	Version     int // Versions should start at 1
+	Action      Action
+}
+
+// Action is something that needs to be done
+type Action interface {
+	Run(log *zap.Logger, db DB, tx *sql.Tx) error
+}
+
+// ValidTableName checks whether the specified table name is valid
+func (migration *Migration) ValidTableName() error {
+	matched, err := regexp.MatchString(`^[a-z_]+$`, migration.Table)
+	if !matched || err != nil {
+		return Error.New("invalid table name: %v", migration.Table)
+	}
+	return nil
+}
+
+// Run runs the migration steps
+func (migration *Migration) Run(log *zap.Logger, db DB) error {
+	err := migration.ValidTableName()
+	if err != nil {
+		return err
+	}
+
+	err = migration.createVersionTable(log, db)
```

Comment:
It might be worth acquiring an advisory lock here, in case it ever happens that two connections try to run migrations at the same time.